README.md
Geobricks Processing
====================

The Geobricks processing library provides an easy way to process sets of layers, providing band extraction and some of the GDAL basic funcionalities. 

Installation
============
The plug-in is distributed through PyPi and can be installed by typing the following command in the console:
```
pip install geobricksprocessing
```
Examples
========

GDALWARP

```python

from geobricks_processing.core import processing_core

processing_gdalwarp = {
    #Mandatory: "Array containing the source paths i.e. a layer of *.tif for band extraction or merging"
    "source_path": ["data/burundi_maize_area_3857.tif"],
    # Mandatory: String containing the output path. If it doesn't exists it will be created
    "output_path": "data/gdalwarp/",
    # Optional: String containing the output file name. If it doesn't exists it will be created with a uuid function
    "output_file_name": "burundi_maize_area_4326.tif",
    # Optional: Default band extraction is 1
    "band": 1,
    # Mandatory: Array with a list of operation (i.e. gdalwarp, gdaladdo, gdal_translate, band_extraction)
    "process": [
        {
            "gdalwarp": {
    # Optional: containing the command options
                "opt": {
                    "-multi": "",
                    "-overwrite": "",
                    "-of": "GTiff",
                    "-s_srs": "EPSG:3857",
                    "-t_srs": "EPSG:4326"
                }
            }
        }
    ]
}

output_files = processing_core.process_data(processing_gdalwarp)
```

GDALWARP with a GDALADDO operation

To do more than one operation can be created an array containing the different steps of the process. In this example will be applied first a GDALWARP and then a GDALADDO operation to the output file.

```python
from geobricks_processing.core import processing_core

processing_gdalwarp = [
    {
        "source_path": ["data/burundi_maize_area_3857.tif"],
        "output_path": "data/gdalwarp/",
        "output_file_name": "burundi_maize_area_4326.tif",
        "band": 1,
        "process": [
            {
                "gdalwarp": {
                    "opt": {
                        "-multi": "",
                        "-overwrite": "",
                        "-of": "GTiff",
                        "-s_srs": "EPSG:3857",
                        "-t_srs": "EPSG:4326"
                    }
                }
            }
        ]
    },
    {
        "band": 1,
        "process": [
            {
                "gdaladdo": {
                    "parameters": {
                        # "--config": "BIGTIFF_OVERVIEW IF_NEEDED"
                    },
                    "overviews_levels": "2 4 8 16"
                }
            }
        ]
    }
]

output_files = processing_core.process_data(processing_gdalwarp)
```




CHANGES.txt
v0.0.1, 24-11-2014 -- Initial release.
v0.0.2, 25-11-2014 -- Fixes.
v0.0.3, 27-12-2014 -- Fixes.
v0.0.4, 16-03-2015 -- Fixes.


test/geobricks_processing_test.py
import unittest
from shutil import rmtree
import os
from geobricks_processing.core import processing_core

path = "../test_data/burundi_maize_area"
source_path = [path + "/burundi_maize_area_3857.tif"]
output_path = path + "/gdalwarp"
if os.path.isdir(output_path):
    rmtree(output_path)

processing_gdalwarp = [
    {
        "source_path": source_path,
        "output_path": output_path,
        "output_file_name": "burundi_maize_area_4326.tif",
        "band": 1,
        "process": [
            {
                "gdalwarp": {
                    "opt": {
                        "-multi": "",
                        "-overwrite": "",
                        "-of": "GTiff",
                        "-s_srs": "EPSG:3857",
                        "-t_srs": "EPSG:4326"
                    }
                }
            }
        ]
    },
    {
        "band": 1,
        "process": [
            {
                "gdaladdo": {
                    "parameters": {
                        # "--config": "BIGTIFF_OVERVIEW IF_NEEDED"
                    },
                    "overviews_levels": "2 4 8 16"
                }
            }
        ]
    }
]


class GeobricksProcessingTest(unittest.TestCase):

    def test_processing_gdalwarp_gdaladdo(self):
        outputfiles = processing_core.process_data(processing_gdalwarp)
        self.assertEqual(len(outputfiles), 1)

geobricks_processing/core/processing_core.py
from osgeo import gdal
import os
import subprocess
import json
import logging
import uuid
from geobricks_common.core.log import logger
from geobricks_common.core.filesystem import tmp_folder

log = logger(__file__)

# methods not related to a GDAL function
key_function = [
    "extract_bands",  # used to extract the bands
    "get_pixel_size"  # used to get the pixel size
]


def process_data(objs, loglevel=logging.INFO):
    """
    Process a json array with GDAL raster operations.
    @return: An array with the processed files.
    """
    result = []
    for obj in objs:
        obj["source_path"] = obj["source_path"] if "source_path" in obj else result
        result = process_obj(obj, loglevel)
    return result


def process_obj(obj, loglevel=logging.INFO):
    """
    Process a json object with GDAL raster operations.
    @return: An array with the processed files.
    """
    log.setLevel(loglevel)
    try:
        source_path = obj["source_path"]
        process = obj["process"]
    except Exception:
        pass
        log.error("Raise exception: output_path, source_path and process type are mandatory")
    output_path = obj["output_path"] if "output_path" in obj else tmp_folder
    output_file_name = obj["output_file_name"] if "output_file_name" in obj else "layer_tmp_" + str(uuid.uuid4())

    if "tmp" in obj:
        output_path = tmp_folder
        output_file_name += str(uuid.uuid4())


    if output_path is not None and not os.path.isdir(output_path):
        os.makedirs(output_path)
    band = obj["band"] if "band" in obj else 1

    # deal with pixel size
    pixel_size = None

    # Processing
    p = Process(output_file_name)

    # default init is the source_path
    output_processed_files = source_path
    # looping throught processes
    for process_values in process:
        for key in process_values:
            if key in key_function:
                # explicit functions
                if "extract_bands" in key:
                    output_processed_files = p.extract_bands(output_processed_files, band, output_path)
                # get the pixel size
                elif "get_pixel_size" in key:
                    pixel_size = p.get_pixel_size(output_processed_files[0], process_values[key])
                    log.info(pixel_size)
            else:
                # Standard GDAL functions
                process_values[key] = change_values(process_values[key], pixel_size)
                # reflection methods calls
                output_processed_files = getattr(p, key)(process_values[key], output_processed_files, output_path)
    return output_processed_files


def change_values(obj, pixel_size):
    s = json.dumps(obj)
    s = s.replace("{{PIXEL_SIZE}}", str(pixel_size))
    return json.loads(s)


# Class to process using reflection
class Process:

    def __init__(self, output_file_name=None):
        if output_file_name is not None:
            self.output_file_name = output_file_name

    def extract_bands(self, input_files, band, output_path):
        file_bands = []
        filenames = []
        ext = None
        try:
            files = input_files
            for f in files:
                gtif = gdal.Open(f)
                sds = gtif.GetSubDatasets()
                file_bands.append(sds[int(band) - 1][0])
                filenames.append(get_filename(f))
                if ext is None:
                    filename, ext = os.path.splitext(f)
            return self.extract_band_files(file_bands, filenames, output_path, ext)
        except Exception, e:
            log.error(e)
            raise Exception(e, 400)

    def extract_band_files(self, input_file_bands, input_filenames, output_path, ext=None):
        output_files = []
        i = 0
        try:
            for file_band, filename in zip(input_file_bands, input_filenames):
                output_file_path = os.path.join(output_path, filename)
                cmd = "gdal_translate '" + file_band + "' " + output_file_path
                log.info(cmd)
                process = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell=True)
                output, error = process.communicate()
                log.info(output)
                output_files.append(output_file_path)
                i += 1
            return output_files
        except Exception, e:
            log.error(e)
            raise Exception(e, 400)

    def get_pixel_size(self, input_file, formula=None):
        # TODO: get pixel value with rasterio library?
        cmd = "gdalinfo " + input_file + " | grep Pixel"
        log.info(cmd)
        try:
            process = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell=True)
            output, error = process.communicate()
            log.info(output)
            if "Pixel Size" in output:
                pixel_size = output[output.find("(")+1:output.find(",")]
                log.info(pixel_size)
                formula = formula.replace("{{PIXEL_SIZE}}", str(pixel_size))
                log.info(formula)
                return eval(formula)
            return None
        except Exception, e:
            log.error(e)
            raise Exception(e, 400)

    def gdal_merge(self, parameters, input_files, output_path):
        output_files = []
        output_file = os.path.join(output_path, self.output_file_name)
        output_files.append(output_file)
        cmd = "gdal_merge.py "
        if "opt" in parameters:
            for key in parameters["opt"].keys():
                cmd += " " + key + " " + str(parameters["opt"][key])
        for input_file in input_files:
            cmd += " " + input_file
        cmd += " -o " + output_file
        log.info(cmd)
        try:
            process = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell=True)
            output, error = process.communicate()
            log.info(output)
            return output_files
        except Exception, e:
            log.error(e)
            raise Exception(e, 400)

    def gdalwarp(self, parameters, input_files, output_path):
        output_files = []
        output_file = os.path.join(output_path, self.output_file_name)
        output_files.append(output_file)
        cmd = "gdalwarp "
        if "opt" in parameters:
            for key in parameters["opt"].keys():
                cmd += " " + key + " " + str(parameters["opt"][key])
        for input_file in input_files:
            cmd += " " + str(input_file)
        cmd += " " + output_file
        log.info(cmd)
        try:
            process = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell=True)
            output, error = process.communicate()
            log.info(output)
            return output_files
        except Exception, e:
            log.error(e)
            raise Exception(e, 400)

    def gdal_translate(self, parameters, input_files, output_path):
        output_files = []
        output_file = os.path.join(output_path, self.output_file_name)
        output_files.append(output_file)
        cmd = "gdal_translate "
        if "opt" in parameters:
            for key in parameters["opt"].keys():
                cmd += " " + key + " " + str(parameters["opt"][key])
        for input_file in input_files:
            cmd += " " + input_file
        cmd += " " + output_file
        log.info(cmd)
        try:
            process = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell=True)
            output, error = process.communicate()
            log.info(output)
            return output_files
        except Exception, e:
            log.error(e)
            raise Exception(e, 400)

    def gdaladdo(self, parameters, input_files, output_path=None):
        output_files = []
        cmd = "gdaladdo "
        for key in parameters["parameters"].keys():
            cmd += " " + key + " " + str(parameters["parameters"][key])
        for input_file in input_files:
            cmd += " " + input_file
            output_files.append(input_file)
        cmd += " " + parameters["overviews_levels"]
        log.info(cmd)
        try:
            process = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell=True)
            output, error = process.communicate()
            log.info(output)
            return output_files
        except Exception, e:
            log.error(e)
            raise Exception(e, 400)


def callMethod(o, name, options, input_files):
    getattr(o, name)(options, input_files)


def get_filename(filepath):
    drive, path = os.path.splitdrive(filepath)
    path, filename = os.path.split(path)
    return os.path.splitext(filename)[0]

geobricks_processing/rest/processing_rest.py
import json
from flask import Blueprint
from flask import Response
from flask import request
from flask.ext.cors import cross_origin
from geobricks_processing.config.config import config

from geobricks_processing.core import processing_core as p

app = Blueprint(__name__, __name__)

@app.route('/discovery/')
@cross_origin(origins='*')
def discovery():
    """
    Discovery service available for all Geobricks libraries that describes the plug-in.
    @return: Dictionary containing information about the service.
    """
    out = {
        'name': 'RASTER PROCESSING',
        'description': 'Core functionalities and services to processes raster data.',
        'type': 'RASTER_PROCESSING'
    }
    return Response(json.dumps(out), content_type='application/json; charset=utf-8')


@app.route('/process/', methods=['POST'])
@cross_origin(origins='*', headers=['Content-Type'])
def process_data_obj():
    user_json = request.get_json()
    out = p.process_data(user_json, config['settings']['logging']['level'])
    return Response(json.dumps(out), content_type='application/json; charset=utf-8')

examples/processing_modis.py
from shutil import rmtree
import os
from geobricks_processing.core import processing_core

path = "../test_data/MODIS/MOD13A2"
source_path = [path + "/*.hdf"]
output_path = path + "/output"
if os.path.isdir(output_path):
    rmtree(output_path)

process_obj = [
    # extract of the band
    {
        "band": 1,
        "source_path": source_path,
        "output_path": output_path + "/extract_bands",
        "process": [
            {
                "extract_bands": ""
            }
        ]
    },
    # Merging of processed bands
    {
        "output_path": output_path + "/merged_bands",
        "output_file_name": "MOD13A2_merged.hdf",
        "process": [
            {
                "gdal_merge": {
                    "opt": {
                        "-n": "-3000",
                        "-a_nodata": "-3000"
                    }
                }
            }
        ]
    },
    {
        "output_path": output_path + "/gdalwarp",
        "output_file_name": "MOD13A2_3857.tif",
        "process": [
            {
                # 4326
                #"get_pixel_size": "{{PIXEL_SIZE}}/111195.496749"
                # 3857
                "get_pixel_size": "{{PIXEL_SIZE}}/2.00000000012"
            },
            {
                "gdalwarp": {
                    "opt": {
                        "-multi": "",
                        "-overwrite": "",
                        "-of": "GTiff",
                        "-s_srs": "'+proj=sinu +R=6371007.181 +nadgrids=@null +wktext'",
                        #"-t_srs": "EPSG:4326",
                        "-t_srs": "EPSG:3857",
                        "-tr": "{{PIXEL_SIZE}} -{{PIXEL_SIZE}}",
                        "-srcnodata": -3000,
                        "-dstnodata": -3000
                    }
                }
            }
        ]
    },
    {
        "output_path": output_path + "/gdal_translate",
        "output_file_name": "MOD13A2_3857.tif",
        "process": [
            {
                "gdal_translate": {
                    "opt": {
                        "-co": "'TILED=YES'",
                        "-co": "'COMPRESS=DEFLATE'",
                        "-a_nodata": -3000
                    }
                }
            }
        ]
    },
    {
        "process": [
            {
                "gdaladdo": {
                    "parameters": {
                        # "-r": "average"
                    },
                    "overviews_levels": "2 4 8 16"
                }
            }
        ]
    }
]

outputfiles = processing_core.process_data(process_obj)